src/kaleidoscope/hid.cpp

```diff
@@ -68,20 +141,16 @@ void pressRawKey(Key mappedKey) {
 }
 
 void pressKey(Key mappedKey) {
-  if (mappedKey.flags & SHIFT_HELD) {
-    pressRawKey(Key_LeftShift);
-  }
-  if (mappedKey.flags & CTRL_HELD) {
-    pressRawKey(Key_LeftControl);
-  }
-  if (mappedKey.flags & LALT_HELD) {
-    pressRawKey(Key_LeftAlt);
-  }
-  if (mappedKey.flags & RALT_HELD) {
-    pressRawKey(Key_RightAlt);
-  }
-  if (mappedKey.flags & GUI_HELD) {
-    pressRawKey(Key_LeftGui);
+  // If the key we're processing is a "pure" modifier key, its modifier flags are added
+  // directly to the report along with the modifier from its keycode byte. We assume that
+  // the flagged modifiers are intended to also modify other keys pressed while this key
+  // is held, so they are allowed unconditionally. If it's not a "pure" modifier key, we
+  // add them to current_mod_flags, and only allow them to affect the report if the latest
+  // keypress includes those modifiers.
+  if (isPureModifier(mappedKey)) {
```

Comment:
By "pure" do you mean a key that has modifier flags on it, but does not have another keycode?

Reply:
After staring at this for a while, I think I might understand what's going on. I'm going to try to restate what you're doing. I'd love to know if I'm anywhere close to the mark.

As we iterate through the keys on the keyboard, 

If a key is toggled on, record what modifier bits it asks for.

If a key is toggled on OR is still on from a previous scan and has modifier bits, but no other functionality, record what modifier bits it asks for.

Once we've finished the scan, if a key was toggled on, strip out any modifier bits that were not either attached to that key or attached to a key that had only modifier bits.

Is that right?

Reply:
By "pure modifier" I mean "a key that only adds modifiers to the report".

During the scan, when processing a key, if its keycode is a modifier, I add both that keycode and all of its modifier flags (if any) to the report immediately (like what Kaleidoscope currently does with all keys), but if the keycode is something else, I add its modifier flags to a separate flags byte (named `current_mod_flags` in this version), but don't put them directly in the report.

When a key toggles on, I record its modifier flags in a third flags byte (`mod_flags_allowed`).

Just before the report is sent, I add whatever modifier flags were recorded, but only the ones currently allowed by the last key pressed, so any held key gets its mod flags masked if those flags bits aren't also contained in whatever key was last pressed.